drivers/flash/flash_cadence_nand.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2023, Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT cdns_nand
+
+#include "socfpga_system_manager.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/kernel.h>
+
+/* Check if reset property is defined */
+#define CDNS_NAND_RESET_SUPPORT DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+
+#if CDNS_NAND_RESET_SUPPORT
+#include <zephyr/drivers/reset.h>
+#endif
+
+#include "flash_cadence_nand_ll.h"
+
+#define DEV_CFG(_dev)  ((const struct flash_cadence_nand_config *)(_dev)->config)
+#define DEV_DATA(_dev) ((struct flash_cadence_nand_data *const)(_dev)->data)
+
+#define FLASH_WRITE_SIZE DT_PROP(DT_INST(0, DT_DRV_COMPAT), block_size)
+
+#ifdef CONFIG_BOARD_INTEL_SOCFPGA_AGILEX5_SOCDK
+#define DFI_CFG_OFFSET 0xFC
+/* To check the DFI register setting for NAND in the System Manager */
+#define DFI_SEL_CHK    (SOCFPGA_SYSMGR_REG_BASE + DFI_CFG_OFFSET)
+#endif
+
+LOG_MODULE_REGISTER(flash_cdns_nand, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_cadence_nand_data {
+	DEVICE_MMIO_NAMED_RAM(nand_reg);
+	DEVICE_MMIO_NAMED_RAM(sdma);
+	/* device info structure */
+	struct cadence_nand_params params;
+	/* Mutex to prevent multiple processes from accessing the same driver api */
+	struct k_mutex nand_mutex;
+#if CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	/* Semaphore to send a signal from an interrupt handler to a thread  */
+	struct k_sem interrupt_sem;
+#endif
+};
+
+struct flash_cadence_nand_config {
+	DEVICE_MMIO_NAMED_ROM(nand_reg);
+	DEVICE_MMIO_NAMED_ROM(sdma);
+#if CDNS_NAND_RESET_SUPPORT
+	/* Reset controller device configuration for Nand*/
+	const struct reset_dt_spec reset;
+	/* Reset controller device configuration for Combo Phy*/
+	const struct reset_dt_spec combo_phy_reset;
+#endif
+#if CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	void (*irq_config)(void);
+#endif
+};
+
+static const struct flash_parameters flash_cdns_parameters = {.write_block_size = FLASH_WRITE_SIZE,
+							      .erase_value = 0xFF};
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+
+struct flash_pages_layout flash_cdns_pages_layout;
+
+void flash_cdns_page_layout(const struct device *nand_dev, const struct flash_pages_layout **layout,
+			    size_t *layout_size)
+{
+	struct flash_cadence_nand_data *const nand_data = DEV_DATA(nand_dev);
+	struct cadence_nand_params *nand_param = &nand_data->params;
+
+	flash_cdns_pages_layout.pages_count =
+		(nand_param->npages_per_block * nand_param->nblocks_per_lun * nand_param->nluns);
+	flash_cdns_pages_layout.pages_size = nand_param->page_size;
+	*layout = &flash_cdns_pages_layout;
+	*layout_size = 1;
```

Comment:
It does not seem you are querying this information from external device, so I guess you could just pre-calculate this at init, instead of doing that for each layout request.